### PR TITLE
ASTPass options

### PR DESCRIPTION
**Pull request by Sytten**
This is a work in progress. 
I can't get the bench to work with instructions count since I don't have a linux workstation.

Related issue: #3169

### CODE REVIEW

**Comment by Sytten** [comment]
I added a test but it then broke the example of embedded connection, so unsure how to deal with that but I think a test with an attached DB is warranted if we move forward.

**Review by weiznich — COMMENTED** [review]
Thanks for your work here :+1:. I must say I'm quite surprised how well this works out. 
I've left a few small comments, otherwise this already looks good for me (assuming that there are no performance regressions.)

I will start a benchmark run on the CI + run the benchmarks tomorrow locally and see what's happen there.

File: diesel/src/query_builder/ast_pass.rs [hunk 1]
```diff
@@ -240,6 +253,22 @@ where
...
+
+    #[cfg_attr(
+        not(feature = "i-implement-a-third-party-backend-and-opt-into-breaking-changes"),
+        doc(hidden)
+    )] // This is used by the `sql_function` macro
+    #[cfg_attr(
+        doc_cfg,
+        doc(cfg(feature = "i-implement-a-third-party-backend-and-opt-into-breaking-changes"))
+    )]
+    pub fn should_skip_from(&self) -> bool {
```
Comment (Sytten) [inline diesel/src/query_builder/ast_pass.rs:265]: Would we prefer to expose the whole options struct instead?
Is the name too vague?

File: diesel/src/query_builder/ast_pass.rs [hunk 2]
```diff
@@ -264,3 +293,12 @@ where
     DebugBinds(&'a mut Vec<&'b dyn fmt::Debug>),
     IsNoop(&'a mut bool),
 }
+
+#[allow(missing_debug_implementations)]
+#[allow(missing_copy_implementations)]
+#[derive(Default)]
+/// This is used to pass down additional settings to the `AstPass`
+/// when rendering the sql string.
+pub struct AstPassToSqlOptions {
```
[the remark this answers is not in the document]
Comment (Sytten) [reply diesel/src/query_builder/ast_pass.rs:302]: We need at least `pub(crate)` since it is used in the query builder (will do that)

**Comment by Sytten** [comment]
Seems like there is an issue with the external database. Not sure how we should deal with that going forward, remembering to attach on all connections is annoying since it is used only for one test. I would be ok with not testing it, but let me know.

**Comment by weiznich** [comment]
> Seems like there is an issue with the external database. Not sure how we should deal with that going forward, remembering to attach on all connections is annoying since it is used only for one test. I would be ok with not testing it, but let me know.

I think it would be easier if that table wouldn't be created by an explicit connection, but only inline within the test by using `diesel::sql_query`. (That's already done for some other tests as well)

**Comment by weiznich** [comment]
As promised here are the benchmark results using instruction count as measurement:
```
bench_trivial_query/diesel/1
                        time:   [7001.9063 cycles 7007.8978 cycles 7016.9640 cycles]
                        change: [-0.2218% -0.0076% +0.2027%] (p = 0.92 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) high mild
  9 (9.00%) high severe
bench_trivial_query/diesel_boxed/1
                        time:   [14527.4865 cycles 14534.4141 cycles 14541.2555 cycles]
                        change: [+13.481% +13.570% +13.653%] (p = 0.00 < 0.05)
                        Performance has regressed.
bench_trivial_query/diesel_queryable_by_name/1
                        time:   [45240.0157 cycles 45318.9432 cycles 45397.0863 cycles]
                        change: [-0.2518% -0.0942% +0.0597%] (p = 0.24 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) low mild
  2 (2.00%) high mild
bench_trivial_query/diesel/10000
                        time:   [24473592.6471 cycles 24473988.4250 cycles 24474410.6667 cycles]
                        change: [-0.0127% -0.0074% -0.0029%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  9 (9.00%) high mild
bench_trivial_query/diesel_boxed/10000
                        time:   [24482378.2459 cycles 24482933.8521 cycles 24483560.7954 cycles]
                        change: [+0.0106% +0.0131% +0.0156%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) high mild
  8 (8.00%) high severe
bench_trivial_query/diesel_queryable_by_name/10000
                        time:   [28641827.1014 cycles 28648860.4708 cycles 28656256.5808 cycles]
                        change: [+0.0426% +0.0745% +0.1056%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

bench_medium_complex_query/diesel/1
                        time:   [24442.7585 cycles 24443.4679 cycles 24444.3294 cycles]
                        change: [+0.2921% +0.3676% +0.4434%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe
bench_medium_complex_query/diesel_boxed/1
                        time:   [44047.5823 cycles 44086.1934 cycles 44126.4421 cycles]
                        change: [+14.236% +14.386% +14.534%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
bench_medium_complex_query/diesel_queryable_by_name/1
                        time:   [136496.4783 cycles 136724.7880 cycles 136938.0960 cycles]
                        change: [-0.6866% -0.5599% -0.4314%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild
bench_medium_complex_query/diesel/10000
                        time:   [25163999.6486 cycles 25166087.0200 cycles 25168230.3291 cycles]
                        change: [-0.0102% +0.0025% +0.0146%] (p = 0.69 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
bench_medium_complex_query/diesel_boxed/10000
                        time:   [25177403.5843 cycles 25179147.4817 cycles 25180956.4327 cycles]
                        change: [-0.0017% +0.0107% +0.0229%] (p = 0.09 > 0.05)
                        No change in performance detected.
bench_medium_complex_query/diesel_queryable_by_name/10000
                        time:   [53713170.0428 cycles 53721886.9414 cycles 53731489.7635 cycles]
                        change: [-0.0103% +0.0130% +0.0363%] (p = 0.28 > 0.05)
                        No change in performance detected.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

Benchmarking bench_loading_associations_sequentially/diesel/bench_loading_associations_sequentially: Collecting 100 samples in estimated 6.3380 s (10k itera                                                                                                                                                            bench_loading_associations_sequentially/diesel/bench_loading_associations_sequentially
                        time:   [4123639.1631 cycles 4124134.9643 cycles 4124568.5638 cycles]
                        change: [+0.0966% +0.1316% +0.1630%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) low severe
  1 (1.00%) low mild

bench_insert/diesel/1   time:   [22855.0073 cycles 22857.8659 cycles 22860.7049 cycles]
                        change: [-1.5976% -1.5812% -1.5646%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  1 (1.00%) high severe
bench_insert/diesel/100 time:   [460524.8167 cycles 460605.9795 cycles 460677.7608 cycles]
                        change: [-0.2800% +0.0391% +0.3456%] (p = 0.83 > 0.05)
                        No change in performance detected.
Found 15 outliers among 100 measurements (15.00%)
  9 (9.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe
```

Baseline is the current master branch. There are 2 regressions (both boxed query benchmarks with 1 element) and 1 improvement.
I assume the regressions happen because we cannot inline stuff for boxed queries and so the compiler cannot optimize away the additional conditions. That written: I'm not sure why insert improves.

**Comment by Sytten** [comment]
@weiznich I did the fixes. I think it should improve performances for other backends since the function will likely be optimized away completely. We might also want to inline the `should_skip_from` function?
Otherwise it would be "ready" so you decide want you want to do with it.

**Comment by Sytten** [comment]
@weiznich I rebased and removed the flag as we discussed. Do you want me to move the PgLikeReturningClause to postgres?

**Review by weiznich — APPROVED** [review]

**Pull request merged**